rollingpin/hostlist.py

```diff
@@ -73,23 +73,33 @@ def resolve_hostlist(host_refs, all_hosts, aliases):
     return resolved_hosts
 
 
-def restrict_hostlist(hosts, start_at, stop_before):
-    if start_at and not any(host.name == start_at for host in hosts):
-        raise HostSelectionError(
-            "--startat: %r not in host list" % start_at)
-
-    if stop_before and not any(host.name == stop_before for host in hosts):
-        raise HostSelectionError(
-            "--stopbefore: %r not in host list" % stop_before)
-
-    if start_at or stop_before:
-        filtered = hosts
-        if stop_before:
-            filtered = itertools.takewhile(
-                lambda host: host.name != stop_before, filtered)
-        if start_at:
-            filtered = itertools.dropwhile(
-                lambda host: host.name != start_at, filtered)
-        return list(filtered)
-    else:
-        return hosts
+def select_canaries(hosts):
+    """Pick representative canary hosts from the full host list.
+
+    The goals are:
+
+    * the canaries should represent all pools in the hostlist
+    * the first host should be from the most common pool
+        * if it's a bad deploy it affects the pool with the most capacity.
+        * hopefully the largest pool will have the most traffic to test on.
+    * the ordering should be stable for repeatability on revert
+
+    To achieve this, we take one host from each pool ordering the pools by
+    descending size and the hosts within a pool by hostname.
```

Comment:
🔕 I wonder if always ordering by the hostname would ever be problematic.  For autoscaled stuff, probably not.  But our job servers could be interesting, since each runs distinctly different code.  I can't think of a better way at the moment though.  It's probably too much of an exception to worry about.

Reply:
that's a good point. i can change it to instance ID which should be safe.

Reply:
done